web/src/markdown.ts

```diff
@@ -454,30 +535,53 @@ function handleEmoji({emoji_name, get_realm_emoji_url, get_emoji_codepoint}) {
     return alt_text;
 }
 
-function handleLinkifier({pattern, matches, get_linkifier_map}) {
-    const {url_template, group_number_to_name} = get_linkifier_map().get(pattern);
+type LinkifierMatch = PrimitiveValue | PrimitiveValue[] | Record<string, PrimitiveValue>;
+
+function handleLinkifier({
+    pattern,
+    matches,
+    get_linkifier_map,
+}: {
+    pattern: RegExp;
+    matches: LinkifierMatch[];
+    get_linkifier_map: GetLinkifierMap;
+}): string {
+    const item = get_linkifier_map().get(pattern);
+    assert(item !== undefined);
+    const {url_template, group_number_to_name} = item;
 
     let current_group = 1;
-    const template_context = {};
+    const template_context: Record<string, LinkifierMatch> = {};
 
     for (const match of matches) {
-        template_context[group_number_to_name[current_group]] = match;
+        const group_name = group_number_to_name[current_group];
+        template_context[group_name] = match;
         current_group += 1;
     }
 
     return url_template.expand(template_context);
 }
 
-function handleTimestamp(time) {
+function handleTimestamp(time: string | number): string {
     let timeobject;
+    let time_string;
+    let time_number;
+    if (typeof time === "string") {
+        time_string = time;
+        time_number = Number(time);
+    } else {
+        time_string = time.toString();
+        time_number = time;
+    }
```

Comment:
This definitely isn't nice but I couldn't think of something nicer

Reply:
Actually I think we can rely on `time` always being a string, given that it's always coming from something like

```js
    if (cap = this.rules.timestamp.exec(src)) {
      src = src.substring(cap[0].length);
      out += this.timestamp(cap[1]);
```